File: go/anyascii.go
package anyascii

import "strings"

func Transliterate(s string) string {
	var sb strings.Builder
	sb.Grow(len(s))
	for _, r := range s {
		sb.WriteString(TransliterateRune(r))
	}
	return sb.String()
}

func TransliterateRune(r rune) string {
	blockNum := uint32(r) >> 8
	block := Block(blockNum)
	lo := 3 * (int(r) & 0xFF)
	if len(block) <= lo {
		return ""
	}
	l := block[lo + 2]
	var len int
	if (l & 0x80) == 0 {
	    len = 3
	} else {
	    len = int(l & 0x7f)
	}
	if len <= 3 {
		return block[lo:lo + len]
	} else {
		i := (int(block[lo]) << 8) | int(block[lo + 1])
		return Strings[i:i + len]
	}
}

File: ruby/lib/any_ascii.rb
require 'zlib'

module AnyAscii

	BLOCKS = Hash.new do |blocks, block_num|
		file_name = File.join(__dir__, 'data', '%03x' % block_num)
		if File.file?(file_name)
			b = IO.binread(file_name)
			zi = Zlib::Inflate.new(-Zlib::MAX_WBITS)
			s = zi.inflate(b)
			zi.close
			block = s.split("\t")
		else
			block = []
		end
		blocks[block_num] = block
	end

	private_constant :BLOCKS

	def self.transliterate(string)
		if string.ascii_only?
			return string
		end
		result = ''
		string.each_codepoint do |cp|
			if cp <= 127
				result << cp
			else
				block_num = cp >> 8
				lo = cp & 0xff
				block = BLOCKS[block_num]
				if block.length > lo
					result << block[lo]
				end
			end
		end
		return result
	end
end
